Source/ProgressionSystemRuntime/Public/Components/PSHUDComponent.h

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Valerii Rotermel & Yevhenii Selivanov
+
+#pragma once
+
+#include "Data/PSTypes.h"
+#include "Data/PSDataAsset.h"
+#include "Components/ActorComponent.h"
+#include "Structures/PlayerTag.h"
+//---
+#include "PSHUDComponent.generated.h"
+
+/**
+ * Implements the core logic on project about Progression System.
+ */
+
+UCLASS(Blueprintable, ClassGroup=(Custom), meta=(BlueprintSpawnableComponent))
+class PROGRESSIONSYSTEMRUNTIME_API UPSHUDComponent final : public UActorComponent
+{
+	GENERATED_BODY()
+
+public:
+	/** Sets default values for this component's properties. */
+	UPSHUDComponent();
+
+	/** Returns the Progression System data asset. */
+	UFUNCTION(BlueprintPure, Category = "C++")
+	static const UPSDataAsset* GetProgressionSystemDataAsset() { return &UPSDataAsset::Get(); }
+
+	/** Returns current saved progression. */
+	UFUNCTION(BlueprintCallable, Category="C++")
+	const FORCEINLINE FPSRowData& GetSavedProgressionRowData() const { return SavedProgressionRowDataInternal; }
+
+	/** Save the progression depends on EEndGameState. */
+	UFUNCTION(BlueprintCallable, Category="C++")
+	void SavePoints(ELevelType Map, FPlayerTag Character, EEndGameState EndGameState);
+
+	/** Returns the endgame reward. */
+	UFUNCTION(BlueprintCallable, Category="C++")
+	int32 GetProgressionReward(ELevelType Map, FPlayerTag Character, EEndGameState EndGameState);
+
+	/** Saves the current progression. */
+	UFUNCTION(BlueprintCallable, Category="C++")
+	void SaveCurrentGameProgression();
+
+	UFUNCTION()
+	void SaveDataAsync();
+
+	/*********************************************************************************************
+	* Protected properties
+	********************************************************************************************* */
+protected:
+	/** Progression System data asset */
+	UPROPERTY(Transient, BlueprintReadOnly, meta = (BlueprintProtected, DisplayName = "Progression System Data Asset"))
+	TSoftObjectPtr<UPSDataAsset> ProgressionSystemDataAssetInternal;
+
+	/** Created Main Menu widget. */
+	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Transient, Category = "C++", meta = (BlueprintProtected, DisplayName = "Progression Menu Widget"))
+	TObjectPtr<class UPSMenuWidget> ProgressionMenuWidgetInternal = nullptr;
```

Comment:
With new UE5.2 Garbage Collection, it's required to nullify manually all runtime (Transient) properties, otherwise it prevents destroying cached object.

Go through all your `Transient` object properties and make sure you null them. For components override `UnregisterComponent()` event, for Widgets is `RemoveFromParent()`, for Subsystem is `Deinitialize()`.

In general, I would suggest, try to cache most of the pointers in your subsystem, so you will nullify all of them in `Deinitialize()` at once

Reply:
done